singularity-eos/eos/eos_base.hpp

```diff
@@ -128,6 +128,13 @@ class EosBase {
   struct is_raw_pointer
       : std::is_same<std::remove_reference_t<std::remove_cv_t<T>>, R *> {};
 
+  // Generic evaluator
+  template <typename Functor_t>
+  PORTABLE_INLINE_FUNCTION void Evaluate(const Functor_t &f) const {
+    CRTP copy = *(static_cast<CRTP const *>(this));
+    f(copy);
```

Comment:
I'll just answer for Josh, since he and I discussed this design before he implemented. Yes--the functor handles everything. I'm imagining a call pattern that looks something like this:

```C++
struct MyPofRhoT {
  MyPofRhoT(Real *rho_, Real *T_, Real *P_, int N_)
    : rho(rho_), T(T_), P(P_), N(N_) {}
  
  template <typename CRTP>
  void operator(CRTP crtp) {
    Kokkos::parallel_for(N, KOKKOS_LAMBDA(int i) {
      P[i] = crtp.PressureFromDensityTemperature(rho[i], T[i], lambda[i]);
    });
  }

  int N;
  Real *rho;
  Real *T;
  Real *P
  LambdaIndexer lambda = NullIndexr; // or something
};
```

And then you could call it like
```C++
eos.Evaluate(MyPofRhoT(rho, T, N));
```

Of course the real benefit is that parallel for can be something else, for example it could be a Kokkos team policy, so that `Evaluate` can be called on device. Or you can do more complex things in the loop body.

Reply:
I think that lives in the lambda, yes.

Reply:
That makes sense.

My line of thought at the time was that if you're chaining together calls you might want to pass a lambda between them.